bevy_lint/src/lints/panicking_methods.rs

```diff
@@ -0,0 +1,290 @@
+//! applications by forcing developers to handle the `Option` or `Result` in their code.
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! #
+//! #[derive(Component)]
+//! struct MyComponent;
+//!
+//! #[derive(Resource)]
+//! struct MyResource;
+//!
+//! fn panicking_query(query: Query<&MyComponent>) {
+//!     let component = query.single();
+//!     // ...
+//! }
+//!
+//! fn panicking_world(world: &mut World) {
+//!     let resource = world.resource::<MyResource>();
+//!     // ...
+//! }
+//! #
+//! # bevy::ecs::system::assert_is_system(panicking_query);
+//! # bevy::ecs::system::assert_is_system(panicking_world);
+//! ```
+//!
+//! Use instead:
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! #
+//! #[derive(Component)]
+//! struct MyComponent;
+//!
+//! #[derive(Resource)]
+//! struct MyResource;
+//!
+//! fn graceful_query(query: Query<&MyComponent>) {
+//!     match query.get_single() {
+//!         Ok(component) => {
+//!             // ...
+//!         }
+//!         Err(error) => {
+//!             error!("Invariant not upheld: {:?}", error);
+//!             return;
+//!         }
+//!     }
+//! }
+//!
+//! fn graceful_world(world: &mut World) {
+//!     let Some(resource) = world.get_resource::<MyResource>() else {
+//!         // Resource may not exist.
+//!         return;
+//!     };
+//!
+//!     // ...
+//! }
+//! #
+//! # bevy::ecs::system::assert_is_system(graceful_query);
+//! # bevy::ecs::system::assert_is_system(graceful_world);
+//! ```
+
+use crate::declare_bevy_lint;
+use clippy_utils::{
+    diagnostics::span_lint_and_help,
+    source::{snippet, snippet_opt},
+    ty::match_type,
+};
+use rustc_hir::{Expr, ExprKind, GenericArgs};
+use rustc_lint::{LateContext, LateLintPass, Lint};
+use rustc_middle::ty::Ty;
+use rustc_session::declare_lint_pass;
+use rustc_span::{Span, Symbol};
+
+declare_bevy_lint! {
+    pub PANICKING_QUERY_METHODS,
+    RESTRICTION,
+    "called a `Query` or `QueryState` method that can panic when a non-panicking alternative exists"
+}
+
+declare_bevy_lint! {
+    pub PANICKING_WORLD_METHODS,
+    RESTRICTION,
+    "called a `World` method that can panic when a non-panicking alternative exists"
+}
+
+declare_lint_pass! {
+    PanickingMethods => [PANICKING_QUERY_METHODS.lint, PANICKING_WORLD_METHODS.lint]
+}
+
+impl<'tcx> LateLintPass<'tcx> for PanickingMethods {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        // Find a method call.
+        if let ExprKind::MethodCall(path, src, args, method_span) = expr.kind {
+            // Get the type of `src` for `src.method()`. We peel all references to that `Foo`,
+            // `&Foo`, `&&Foo`, etc. all look identical, since method calls automatically
+            // dereference the source.
+            let src_ty = cx.typeck_results().expr_ty(src).peel_refs();
+
+            // Check if `src` is a panicking type (e.g. `Query`), else exit.
```

Comment:
Nit: The type isn't really panicking, but rather offers some methods which can panic.
We can probably leave the enum name as-is, but perhaps we can adjust the comment a bit to clarify this